### PR TITLE
[pfc] Add command line to enable/disable/show PFC

### PR DESCRIPTION
**Pull request by baiwei0427**
Signed-off-by: Wei Bai baiwei0427@gmail.com

**- What I did**
Current PFC cli only allows operators to enable/disable per-port asymmetric PFC. I extend add new commands to 1) enable/disable PFC on a given priority of a given port and 2) check PFC configuration.  I also link all these commands to <code>show</code> and <code>config</code> utilities.

**- How I did it**
For a interface <code>"EthernetX"</code>, I read the field <code>"pfc_enable"</code>of entry <code>“PORT_QOS_MAP|EthernetX"</code> from config DB table to get its lossless priorities. To enable or disable a specific priority, we just need to modify the field <code>"pfc_enable"</code>.

**- How to verify it**
Enable or disable a priority of a interface. Then check the corresponding entry in config DB. 

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**
 Show PFC configuration of interface Ethernet0
<pre>
$ pfc show priority Ethernet0

Interface    Lossless priorities
-----------  ---------------------
Ethernet0    3,4
</pre>

<pre>
$ show pfc priority Ethernet0

Interface    Lossless priorities
-----------  ---------------------
Ethernet0    3,4
</pre>
Disable PFC on priority 3 of interface Ethernet0
<pre>
$ pfc config priority off Ethernet0 3

Interface      Lossless priorities
-----------  ---------------------
Ethernet0                        4
</pre>
<pre>
$ config interface pfc priority Ethernet0 3 off

Interface      Lossless priorities
-----------  ---------------------
Ethernet0                        4
</pre>

Disable PFC on priority 8 (invalid value) of interface Ethernet0
<pre>
$ pfc config priority off Ethernet0 8
Usage: pfc config priority [OPTIONS] STATUS INTERFACE PRIORITY

Error: Invalid value for "priority": invalid choice: 8. (choose from 0, 1, 2, 3, 4, 5, 6, 7)
</pre>
<pre>
$ config interface pfc priority Ethernet0 8 off
Usage: config interface pfc priority [OPTIONS] <interface_name> PRIORITY STATUS

Error: Invalid value for "priority": invalid choice: 8. (choose from 0, 1, 2, 3, 4, 5, 6, 7)
</pre>

Enable PFC on priority 3 of interface Ethernet0
<pre>
$ pfc config priority on Ethernet0 3

Interface    Lossless priorities
-----------  ---------------------
Ethernet0    3,4
</pre>
<pre>
$ config interface pfc priority Ethernet0 3 on

Interface    Lossless priorities
-----------  ---------------------
Ethernet0    3,4
</pre>

### CODE REVIEW

**Comment by andriymoroz** [comment]
I would suggest to link it to the `config` and `show` utilities to have unified CLI
i.e.
```
$ show pfc priority Ethernet0

Interface    Lossless priorities
-----------  ---------------------
Ethernet0    3,4
```

And describe new CLI in the Command Reference doc
https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md

**Comment by baiwei0427** [comment]
> I would suggest to link it to the `config` and `show` utilities to have unified CLI
> i.e.
> 
> ```
> $ show pfc priority Ethernet0
> 
> Interface    Lossless priorities
> -----------  ---------------------
> Ethernet0    3,4
> ```
> 
> And describe new CLI in the Command Reference doc
> https://github.com/Azure/sonic-utilities/blob/master/doc/Command-Reference.md

This is a good suggestion. I can update 'show' and 'config' to support following commands:
<pre>
show pfc asymmetric
show pfc asymmetric [interface]
show pfc priority
show pfc priority [interface]
config pfc asymmetric on [interface]
config pfc asymmetric off [interface]
config pfc on [interface] [priority]
config pfc off [interface] [priority]
</pre>

**Comment by baiwei0427** [comment]
@andriymoroz I have linked the commands to show and config utilities and updated Command Reference doc

File: config/main.py [hunk 1]
```diff
@@ -2007,6 +2007,24 @@ def asymmetric(ctx, interface_name, status):
 
     run_command("pfc config asymmetric {0} {1}".format(status, interface_name))
 
+#
+# 'pfc config priority' command
+#
+
+@pfc.command()
+@click.argument('interface_name', metavar='<interface_name>', required=True)
+@click.argument('priority', type=click.Choice(['3', '4']))
```
Comment (wendani) [inline config/main.py]: Looking forward, is there a way not to hard-code the lossless TCs?

**Comment by wendani** [comment]
You may need https://github.com/Azure/sonic-swss/pull/1055 to check if qosorch is responding to your CLI.

File: config/main.py (hunk 1 above)
Comment (baiwei0427) [reply config/main.py]: @wendani How about using 0-7? I think almost all ASICs only support 8 priorities.

Comment (baiwei0427) [reply config/main.py]: @wendani Support PFC on 0-7 now.

File: doc/Command-Reference.md [hunk 2]
```diff
@@ -3559,6 +3587,67 @@ This command displays the details of Rx & Tx priority-flow-control (pfc) for all
...
+  ```
+
+**show pfc priority**
+
+This command displays the lossless priorities for all interfaces or a given interface.
+
+- Usage:
+  ```
+  show pfc priority
+  show pfc priority <interface>
```
Comment (jleveque) [inline doc/Command-Reference.md]: Combine these into one line, indicating that the <interface> parameter is optional, like this:

```
  show pfc priority [<interface>]
```

File: doc/Command-Reference.md [hunk 3]
```diff
@@ -3559,6 +3587,67 @@ This command displays the details of Rx & Tx priority-flow-control (pfc) for all
...
   ```
 
+**show pfc asymmetric**
+
+This command displays the status of asymmetric PFC for all interfaces or a given interface.
+
+- Usage:
+  ```
+  show pfc asymmetric
+  show pfc asymmetric <interface>
```
Comment (jleveque) [inline doc/Command-Reference.md]: Combine these into one line, indicating that the <interface> parameter is optional, like this:

```
  show pfc asymmetric [<interface>]
```

Comment (baiwei0427) [reply doc/Command-Reference.md]: @jleveque Fixed

**Comment by baiwei0427** [comment]
retest this please

File: doc/Command-Reference.md (hunk 2 above)
Comment (baiwei0427) [reply doc/Command-Reference.md]: @jleveque  Fixed

File: config/main.py [hunk 4]
```diff
@@ -1991,7 +1991,7 @@ def pfc(ctx):
 
 
 #
-# 'pfc asymmetric' command
+# 'pfc config asymmetric' command
```
Comment (jleveque) [inline config/main.py]: Change to:

```
# 'pfc asymmetric' command ('config interface pfc asymmetric ...')
```

Also change line 1983 above to:

```
# 'pfc' subgroup ('config interface pfc ...')
```

Also, after this PR is merged, can you open a new PR to move all of these PFC commands to be with the rest of the `config interface ...` commands? At some point, the commands changed from `config pfc ...` to `config interface pfc ...`.

File: config/main.py [hunk 5]
```diff
@@ -2007,6 +2007,24 @@ def asymmetric(ctx, interface_name, status):
 
     run_command("pfc config asymmetric {0} {1}".format(status, interface_name))
 
+#
+# 'pfc config priority' command
```
Comment (jleveque) [inline config/main.py]: Change to:

```
# 'pfc priority' command ('config interface pfc priority ...')
```

File: doc/Command-Reference.md [hunk 6]
```diff
@@ -2379,6 +2379,34 @@ VLAN interface names take the form of `vlan<vlan_id>`. E.g., VLAN 100 will be na
   admin@sonic:~$ sudo config interface vlan100 ip remove 10.11.12.13/24
   ```
 
+**config interface pfc priority <interface_name> PRIORITY STATUS (Versions >= 201904)**
```
Comment (jleveque) [inline doc/Command-Reference.md]: Change to:

```
**config interface pfc priority <interface_name> <priority> (on | off)**
```

Comment (baiwei0427) [reply doc/Command-Reference.md]: @jleveque Fixed

File: config/main.py (hunk 5 above)
Comment (baiwei0427) [reply config/main.py]: @jleveque Fixed

File: config/main.py (hunk 4 above)
Comment (baiwei0427) [reply config/main.py]: @jleveque Fixed

File: doc/Command-Reference.md [hunk 7]
```diff
@@ -2379,6 +2379,34 @@ VLAN interface names take the form of `vlan<vlan_id>`. E.g., VLAN 100 will be na
   admin@sonic:~$ sudo config interface vlan100 ip remove 10.11.12.13/24
   ```
 
+**config interface pfc priority <interface_name> <priority> (on | off) (Versions >= 201904)**
```
Comment (jleveque) [inline doc/Command-Reference.md]: Remove `(Versions >= 201904)`, as this command is not in that version.

Comment (baiwei0427) [reply doc/Command-Reference.md]: @jleveque Fixed

File: doc/Command-Reference.md [hunk 8]
```diff
@@ -3559,6 +3587,65 @@ This command displays the details of Rx & Tx priority-flow-control (pfc) for all
   root@sonic:~# sonic-clear pfccounters
   ```
 
+**show pfc asymmetric**
```
Comment (volodymyrsamotiy) [inline doc/Command-Reference.md]: This looks like duplication, we already have Asymmetric PFC mode shown as part of ```show interfaces status``` command (it is also described in ```Command-Reference.md``` doc under the ```show interfaces status``` section):
```
root@sonic:/home/admin# show interfaces status
      Interface            Lanes    Speed    MTU    Alias             Vlan    Oper    Admin             Type    Asym PFC
---------------  ---------------  -------  -----  -------  ---------------  ------  -------  ---------------  ----------
      Ethernet0          0,1,2,3      40G   9100     etp1           routed    down     down   QSFP+ or later         off
      Ethernet4          4,5,6,7      40G   9100     etp2            trunk      up       up   QSFP+ or later         off
...
```

Comment (baiwei0427) [reply doc/Command-Reference.md]: @volodymyrsamotiy Yes, this is true. But there is no harm.

**Pull request merged**